gipc/gipc.py

```diff
@@ -327,21 +340,21 @@ def _child(target, args, kwargs):
         # could even reset sigprocmask (Python 2.x does not have API for it, but
         # it could be done via ctypes).
         _reset_signal_handlers()
+
         # `gevent.reinit` calls `libev.ev_loop_fork()`, which reinitialises
         # the kernel state for backends that have one. Must be called in the
         # child before using further libev API.
-        gevent.reinit()
-        log.debug("Delete current hub's threadpool.")
         hub = gevent.get_hub()
-        # Delete threadpool before hub destruction, otherwise `hub.destroy()`
-        # might block forever upon `ThreadPool.kill()` as of gevent 1.0rc2.
-        del hub.threadpool
-        hub._threadpool = None
-        # Destroy default event loop via `libev.ev_loop_destroy()` and delete
-        # hub. This orphans all registered events and greenlets that have been
-        # duplicated from the parent via fork().
-        log.debug("Destroy hub and default loop.")
-        hub.destroy(destroy_loop=True)
+        
+        # Destroy the loop, to prevent any old greenlets from processing
+        hub.loop.destroy()
+        hub.loop = None
```

Comment:
This diff raises two interesting questions:

- what is the key difference between `hub.destroy(destroy_loop=True)` and `hub.loop.destroy()`?
- does the order matter between `gevent.reinit(); hub.loop.destroy()` or `hub.loop.destroy(); gevent.reinit()`? if yes: why?

You can see that above I commented

> # `gevent.reinit` calls `libev.ev_loop_fork()`, which reinitialises  
>  # the kernel state for backends that have one

is that still true?

Reply:
> what is the key difference between hub.destroy(destroy_loop=True) and hub.loop.destroy()?

I think we have found that this is the essence of this change here -- I have added a commit to #109 with that minimal change, and a code comment summarizing my current thinking about this: https://github.com/jgehrcke/gipc/pull/109/commits/4a79bbcfd8401d2381b423af4304568da1c570df

Reply:
Pretty sure calling reinit first restarts the loop on the forked process and begins executing greenlets, that's why I execute the destroy first.

I'm not 100% on why, but there seems to be a behavior change in libev on fork that unsets the loop _ptr, which then causes _check_loop to fail when hub.destroy tries to throw the HubDestroyed exception on all the greenlets

Reply:
I pulled it in, and it appears not to... so I suppose it's probably better to do it first.

I should also note that if gevent is monkeypatched, reinit appears to get called automatically on fork... If gevent is not monkey patched, then it does not get called automatically.

I would imagine multiple calls to reinit won't be a problem

Reply:
Thanks for the feedback here. Contributed to both our understanding :).